flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/FineGrainedTaskManagerTracker.java

```diff
@@ -118,6 +124,34 @@ public void removeTaskManager(InstanceID instanceId) {
         }
     }
 
+    @Override
+    public void addUnWantedTaskManager(InstanceID instanceId) {
+        TaskManagerInfo taskManagerInfo = taskManagerRegistrations.get(instanceId);
+        if (taskManagerInfo != null) {
+            addUnWantedTaskManagerWithResource(
+                    instanceId,
+                    taskManagerInfo.getTotalResource(),
+                    taskManagerInfo.getDefaultSlotResourceProfile());
+        }
+    }
+
+    public void addUnWantedTaskManagerWithResource(
+            InstanceID instanceId,
+            ResourceProfile totalResourceProfile,
+            ResourceProfile defaultSlotResourceProfile) {
+        unWantedTaskManagers.put(
+                instanceId,
+                WorkerResourceSpec.fromTotalResourceProfile(
+                        totalResourceProfile,
+                        SlotManagerUtils.calculateDefaultNumSlots(
+                                totalResourceProfile, defaultSlotResourceProfile)));
+    }
```

Comment:
Again, why do we need 2 interfaces here for adding unwanted workers?

Reply:
I think the underlying question is that, what is the relation ship between `unWantedTaskManagers` and `taskManagerRegistrations`? Is unwanted workers always a subset of registered workers?